### ISSUE TITLE
Downloads from Chocolatey.org returning bad files and other site issues

### ISSUE THREAD

**Issue by ferventcoder** [chocolatey/chocolatey.org#236]
tl;dr - **On October 23, 2015, at just before 17:00 CDT we implemented a new caching strategy and increased the timeout to nearly remove this from happening in the future.**

For the better part of October, Chocolatey.org was under an extreme traffic load and the database started timing out and returning 404 part of the time. 

In some cases that meant that package requests would come back as 5KB files that were really the 404 HTML page and the same for the installer zip. The zip or nupkg will be about 5KB and if you rename it to .html you will see the error page html in a text editor. Here's a great example of what this looks like: https://github.com/chocolatey/chocolatey/issues/710

## CDN Implemented October 10, 2015
On October 10 at 09:40 CDT, we introduced CloudFlare as a CDN in front of Chocolatey.org and that relieved some of the pressure - https://twitter.com/chocolateynuget/status/652896815679860736. However there were some issues related to package downloads so we backed off on that aspect until we could reintroduce it for the packages themselves, which was done on October 12, 2015 at 23:00 CDT - https://twitter.com/chocolateynuget/status/653792658628804609.

To give some analytics about what CloudFlare has provided, the bandwidth is right around 1TB per day:
![image](https://cloud.githubusercontent.com/assets/63502/10711017/90e3253a-7a34-11e5-8c56-912c0cb7fc7d.png)

The number of requests per day is somewhere between 5 and 6 Million (it tends to go down a bit on the weekends):
![image](https://cloud.githubusercontent.com/assets/63502/10711023/c8cc6704-7a34-11e5-8f46-0e8422edbf64.png)

## New Caching Strategies Implemented October 23, 2015
After the CDN implementation it became clear that it was not enough. Starting on Wednesday, October 21, 2015, users were reporting time outs and other issues surrounding trying to install Chocolatey itself, which uses a script from the website. The script was fine, but the request to download the package was pulling down a zip file that was essentially a 404 HTML file due to time outs on the site and the underlying database.

Knowing that there is nearly zero caching implemented for most of the calls to Chocolatey.org, we implemented better and heavier caching techniques, introducing Glav.CacheAdapter after having good experiences doing the same on other sites. We also increased the command timeout for queries to the database. The work was done in https://github.com/chocolatey/chocolatey.org/pull/232. In adding better caching techniques, it became clear that it was going to conflict with the current indexing methods used on the site so that was removed. That should also fix other things like seeing packages in search results sooner after updates. There are 3 currently known issues:

* Package Upload from site may not work - https://github.com/chocolatey/chocolatey.org/issues/233
* Search is case sensitive - https://github.com/chocolatey/chocolatey.org/issues/234
* Search results returning by popularity, not by relevance - https://github.com/chocolatey/chocolatey.org/issues/235

This new caching strategy was introduced on October 23, 20015 just before 17:00 CDT. It's perhaps a little early to tell whether additional work needs to be done given weekends are typically lower traffic.

Here are some early graphs of the results.

##### Http Response Time over 7 days (lower is better)
![image](https://cloud.githubusercontent.com/assets/63502/10711790/798cdce2-7a4c-11e5-8b00-57973c674890.png)

##### Active Requests over 7 days
![image](https://cloud.githubusercontent.com/assets/63502/10711084/79e13ae6-7a36-11e5-8215-2af906db9dff.png)

##### Load Average over 7 days (lower is better)
![image](https://cloud.githubusercontent.com/assets/63502/10711234/9b4b3494-7a3a-11e5-894f-bece1d7c9999.png)

##### Process Handles over 7 days (lower is better)
![image](https://cloud.githubusercontent.com/assets/63502/10711423/4fcde966-7a40-11e5-92ad-5de30a09f42e.png)

##### Memory Usage over 7 days (lower is better)
This is where we expected an increase. We'll be following up with something like memcached here to remove the load from each individual server. 
![image](https://cloud.githubusercontent.com/assets/63502/10711263/3d8e5272-7a3b-11e5-909f-b56c57c16841.png)

**Comment by ferventcoder** [comment]
Here's an example of the behavior that was being seen - https://github.com/chocolatey/chocolatey/issues/710

**Comment by jimaek** [comment]
@ferventcoder Great job on the caching! 
Send me an email if you need a free CDN for any part of Chocolatey including the distribution.

**Comment by ferventcoder** [comment]
Adding more charts here over the next hour.

**Comment by ferventcoder** [comment]
![image](https://cloud.githubusercontent.com/assets/63502/10863989/b1e49660-7fac-11e5-8802-fe9db4cc9ea0.png)
